# Table of Contents

01. Import Libraries
02. Import Data Sets
03. Practicing fix of mixed-type columns
04. Missing values - Finding them...
05. Duplicates - Finding them...
06. Exporting df_prods_clean_no_dups
07. Task 4.5 - Answers

# 01. Import Libraries

In [1]:
# Import pandas, numpys and os

import pandas as pd
import numpy as np
import os

# 02. Import Data Sets

In [2]:
# create path variable

path = r'/Users/dianaalatriste/Documents/Instacart'

In [3]:
path

'/Users/dianaalatriste/Documents/Instacart'

In [4]:
# Import products.csv data set

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import orders_wrangled.csv data set

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Practicing fix of mixed-type columns

In [6]:
# Create a new test dataframe

df_test = pd.DataFrame()

In [7]:
# Create a mixed type column within new dataframe

df_test['mix'] = ['a', 'b', 1, True]

In [8]:
# Checking for new data frame

df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types in df_test

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [12]:
#Checking for data type of column 'mix'

df_test.dtypes

mix    object
dtype: object

In [13]:
# Changing data type of column mix

df_test['mix'] = df_test['mix'].astype('str')

# 04. Missing values - Finding them...

In [14]:
# Find missing values in df_prods

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# Create a new dataframe that only includes the null values from the 'product_name' column. Dataframe name = df_nan

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Missing values can't be imputed because they are string values.

In [17]:
# Checking number of rows in current dataframe to compare with new dataframe once missing values have been removed

df_prods.shape

(49693, 5)

## Creating a new dataframe without missing values called df_prods_clean

In [18]:
#Creating new dataframe without missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
# Checking number of rows in df_prods_clean

df_prods_clean.shape

(49677, 5)

#### New dataframe has exactly 16 values less than the original one. These are the missing values. 

# 05. Duplicates - Finding them...

In [20]:
# Create a new subset from df_prods_clean that contains only full duplicates (aka rows that have the exact same values in every column)

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

## Addressing duplicates

In [22]:
# Check first for number of rows in df_prods_clean

df_prods_clean.shape

(49677, 5)

In [23]:
# Create a new dataframe that does not include duplicates using .drop_duplicates function

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

#### New dataframe df_prods_clean_no_dups contains 5 less rows, the exact number of duplicates

# 06. Exporting df_prods_clean_no_dups

In [31]:
#Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5 - Answers

## Task 4.5.2. Run df.describe() function on your df_ords dataframe.

In [27]:
# Running describe() on df_ords. 

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  last_purchase  
count        3.421083e+06       3.421083e+06   3.214874e+06  
mean         2.776219e+00       1.345202e+01   1.111484e+01  
std          2.046829e+00       4.226088e+00   9.206737e+00  
min          0.000000e+00       0.000000e+00   0.000000e+00  
25%          1.000000e+00       1.000000e+01   4.000000e+00  
50%          3.000000e+00       1.300000e+01   7.000000e+00  
75%          5.000000e+00       1.600000e+01   1.500000e+01  
max          6.000000e+00       2.300000e+01   3.000000e+01

In [44]:
# Dropping 'Unnamed:0' column in df_ords

df_ords = df_ords.drop('Unnamed: 0', axis = 1)

AttributeError: 'NoneType' object has no attribute 'drop'

In [47]:
# reimporting orders_wrangled. csv as df_ords

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

In [48]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  last_purchase  
0                  8            NaN  
1                  7           15.0  
2                 12           21.0  
3                  7           29.0  
4                 15           28.0

In [49]:
# Rerunning describe() on df_ords. 

df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  last_purchase  
count       3.421083e+06   3.214874e+06  
mean        1.345202e+01   1.111484e+01  
std         4.226088e+00   9.206737e+00  
min         0.000000e+00   0.000000e+00  
25%         1.000000e+01   4.000000e+00  
50%         1.300000e+01   7.000000e+00  
75%         1.600000e+01   1.500000e+01  
max         2.300000e+01   3.000000e+01

#### I don't seem to spot any difference, particularly when focusing on minimim and maximum values of 'orders_day_of_week' , 'order_hour_last_day' and 'last_purchase'

## Task 4.5.3. Check for mixed-type data in your df_ords dataframe.

In [59]:
# Check for mixed types in df_ords

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)
print('Your code has been run')

Your code has been run


### - There are no mixed-type data in df_ords dataframe

## Task 4.5.4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

### - No column to fix

## Task 4.5.5. Run a check for missing values in your df_ords dataframe.

In [53]:
# Find missing values in df_ords

df_ords.isnull().sum()

order_id                   0
user_id                    0
eval_set                   0
order_number               0
orders_day_of_week         0
order_hour_of_day          0
last_purchase         206209
dtype: int64

In [54]:
# Check shape of df_ords

df_ords.shape

(3421083, 7)

#### There are 206209 missing values in columns 'last_purchase', formerly knows as 'days_since_prior_order'. My guess is that these users have created an account at Instacart but have only made a one purchase since the creation of their account. Therefore there is no data regarding their last purchase. 

## Task 4.5.6. Address the missing values using an appropriate method

### The missing data is valuable as it points out to new costumers which can be valuable for further analysis. Missing values can't really be ignored as they represent only the 6.02% of total users and it can't be erased as then this information will be lost. One solution would be to create a subset of new customers/ customers with just one purchase. However, I will not do it this time and decide if this is worth it later in the course. 

# Task 4.5.7. Run a check for duplicate values in your df_ords data.

In [58]:
# Checking for duplicates in df_ords

df_ords_dups = df_ords[df_ords.duplicated()]
print('Your code has been run')

Your code has been run


#### There were no duplicates identified

In [2]:
# Check output df_ords_dups

df_ords_dups.head()

NameError: name 'df_ords_dups' is not defined

# Task 4.5.8. Address the duplicates using an appropriate method

#### No duplicates were found therefore there's nothing to be addressed.

# Task 4.5.9. Exporting df_ords as .csv file in 'Prepared_Data' folder

In [60]:
# Export df_ords to 'Prepared_Data folder'

df_ords_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [61]:
print ('End of Task')

End of Task
